### PR TITLE
Add Support for Embedding Tasks

### PR DESCRIPTION
**Pull request by drewrip**
## Description
This PR will add support in the client library for [the embedding tasks in caikit](https://github.com/markstur/caikit-embeddings).

## How Has This Been Tested?
For now, this has been tested locally using some examples I have, and an embedding model I have deployed via RHOAI. The formal testing will have to be updated to support these new endpoints as well.

## Merge criteria:

- [ ] The commits are squashed in a cohesive manner and have meaningful messages.
- [ ] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has manually tested the changes and verified that the changes work

## Note

I needed to be able to query the embeddings endpoints via Python anyway, so I figured I'd add them here.  The exact signatures I added might not be what is desired in the final client library but for my purposes it is more helpful to expose all the details of the responses, including the number of tokens that were provided as input. I need that metric for my testing.

This is also only the HTTP client for now. I will try to add that if I have time later, or it sounds like @dtrifiro might already have code for this

### CODE REVIEW

File: src/caikit_nlp_client/http_client.py [hunk 1]
```diff
@@ -292,3 +291,180 @@ def _create_json_request(self, model_id, text, **kwargs) -> dict[str, Any]:
...
+        """
+        if response.status_code == 200:
+            return response.json()
+
+        try:
+            details = response.json()[details_field]
+        except json.JSONDecodeError:
+            details = f"{response.reason} {response.text=}"
+
+        raise RuntimeError(f"{response.status_code=} {details}")
```
Comment (dtrifiro) [inline src/caikit_nlp_client/http_client.py]:
```suggestion
        try:
            details = response.json()[details_field]
        except json.JSONDecodeError as exc:
            details = f"{response.reason} {response.text=}"
            raise RuntimeError(f"{response.status_code=} {details}") from exc
```

Comment (drewrip) [reply src/caikit_nlp_client/http_client.py]: Not sure I understand? I tried to use the same pattern of error handling [already present](https://github.com/opendatahub-io/caikit-nlp-client/blob/c97f7d31f58d6feed60be71c09eb0d39c0eb98ef/src/caikit_nlp_client/http_client.py#L183) in the HTTP client

**Comment by drewrip** [comment]
I'm not sure if it is intentional, but i haven't found any docs about this so I opened an issue about the `details` / `detail` issue [here](https://github.com/caikit/caikit/issues/750)

File: src/caikit_nlp_client/http_client.py (hunk 1 above)
Comment (dtrifiro) [reply src/caikit_nlp_client/http_client.py]: Oh well that's fine, I guess we need to clean those up.

File: src/caikit_nlp_client/http_client.py [hunk 2]
```diff
@@ -292,3 +291,188 @@ def _create_json_request(self, model_id, text, **kwargs) -> dict[str, Any]:
...
+        The details_field is present because there seems to be an inconsistent naming of this
+        field between different endpoints. Some have "details" others have "detail".
+
+        When details_field == None, try to detect the right name of the details field
+        """
+        if response.status_code == 200:
+            return response.json()
+
+        try:
+            if details_field is None:
```
Comment (dtrifiro) [inline src/caikit_nlp_client/http_client.py:535]: Can you add a comment pointing to https://github.com/caikit/caikit/issues/750?

Comment (drewrip) [reply src/caikit_nlp_client/http_client.py:535]: Added with [a0f761e](https://github.com/opendatahub-io/caikit-nlp-client/pull/135/commits/a0f761ea9da18dcaf4b52a2e571cb26044f78d59)

**Comment by dtrifiro** [comment]
I took the liberty of rebasing and doing a few changes @drewrip. Are you able to add a few tests for these?

**Comment by drewrip** [comment]
@dtrifiro Looks good. I can probably add some tests and document the new functions next week. I'm currently using this branch for some downstream testing of these endpoints. Any chance someone on the team has worked on the gRPC endpoints? I'm not sure if I'll have time to tackle those for a bit.

**Comment by dtrifiro** [comment]
@drewrip yeah, I have a rough draft sketched out. I can push the wip commits to this branch (or make another one based on this). Let me know what you prefer

**Comment by drewrip** [comment]
@dtrifiro Yeah doing that in another PR is probably a better idea. Let me add the documentation and tests, then I think this will be good from my side :)

**Comment by dtrifiro** [comment]
Here's the branch: https://github.com/dtrifiro/caikit-nlp-client/tree/grpc-client-tasks

**Comment by drewrip** [comment]
@dtrifiro I added some basic tests with [80f1be5](https://github.com/opendatahub-io/caikit-nlp-client/pull/135/commits/80f1be56c50b3748f0d249ef84ec0f1125a1eddc), however none of them are passing. I don't really understand the testing framework so I'm not sure how to fix this, but the model that is being used in the tests is setup for text generation. So the result of the embedding tests is either errors or gibberish

**Comment by dtrifiro** [comment]
Thanks @drewrip, in order to test those endpoints, we need to add an embedding model to the models in `tests/tiny_models`. There's a quick guide here: https://github.com/opendatahub-io/caikit-nlp-client/blob/main/tests/tiny_models/README.md

I'll shortly push something

**Comment by dtrifiro** [comment]
@drewrip I updated and added an `embedding_model_name` fixture that can be used to use an embedding model.

I also removed the grpc client commits, opting to move them to #140  instead, as that involves a larger refactor. You can check out that PR out. It also has working embedding tests for the grpc client (although with limited functionality testing)

**Comment by drewrip** [comment]
@dtrifiro Gotcha, thanks for making those changes for me! Everything seems to look good from my end then. Let me know if there is anything else needed for this PR that I can contribute at this point

**Comment by dtrifiro** [comment]
That's all good, I'll merge this and then #140 and then we can take care of leftovers. I'll release a new version once we're satisfied with all the changes

**Comment by drewrip** [comment]
Thanks for the help! This will be helpful for some of the testing PSAP is doing :)

**Pull request merged**